dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NormalizeLoops.java

```diff
@@ -69,5 +104,46 @@ public void run(Function function) {
 
             counter++;
         }
+
+        // Guarantees header is the only entry point
+        for (Label label : function.getEvents(Label.class)) {
+            final List<CondJump> backJumps = label.getJumpSet().stream()
+                    .filter(j -> j.getLocalId() > label.getLocalId())
+                    .sorted()
+                    .toList();
+            final Label loopBegin = label;
+            final CondJump uniqueBackJump = backJumps.get(0);
+
+            final List<Label> loopBodyLabels = loopBegin.getSuccessor().getSuccessors().stream()
+                    .takeWhile(ev -> ev != uniqueBackJump)
+                    .filter(Label.class::isInstance).map(Label.class::cast)
+                    .toList();
+            for (Label l : loopBodyLabels) {
+                final List<CondJump> externalEntries = l.getJumpSet().stream()
+                        .filter(j -> j.getLocalId() < loopBegin.getLocalId() ||
+                                j.getLocalId() > uniqueBackJump.getLocalId())
```

Comment:
> If this was ever true, you would have intertwined (non-wellnested) loops. 

What do you mean?

> You should throw on exception or let `LoopFormVerification` throw the exception (if it even checks for such strange loops).

This is exactly the condition the `LoopFormVerification` is checking (and throwing an exception if detected). I copied the condition from there.

Reply:
The highlighted part of my review went missing... I am specifically referring to the second check of your condition: `j.getLocalId() > uniqueBackJump.getLocalId()`.
If this condition is ever true you would have two partially overlapping loops: the backjump of one loop would land in the middle of another loop.
In this case, your code would retarget the backjump of the second loop to go the header of the first loop, essentially merging them into a single loop. However, this merged loop will now have two backjumps, the original one of the first loop and the newly obtained one from the second one.

Reply:
Where is the second loop coming from?

Consider this code
```
e1: header:
...
e2: label:  // in loopBodyLabels
...
e3: goto header // backjump
...
e4: goto label // in externalEntries
```
Here we have `j.getLocalId() > uniqueBackJump.getLocalId()` since e4 > e3, but we don't have two overlapping loops.

Reply:
I see two overlapping loops `(e1, e3)` and `(e2, e4)`. Of course, you can come up with cases where the second loop is not an actual semantic loop (i.e., `e4` is unreachable from `e2`), but that is not clear at all from the syntactic structure.
Either way, even if for some odd reason the second loop is not a semantic loop, the issue still stands: after changing `e4` to go to `e1` the loop gets a new backjump.
This would invalidate the unique back jump pass. If you run the unique backjump pass after doing your transformation, it might not be an issue though.

Reply:
I was under the (wrong) impression that `e4` was not reachable from `e2` since `e3` is unconditional, but in principle we could have a jump to `e4` between `e2` and `e3`.

However, the current code should already take care of this because as you said, the single backjump transformation pass is applied after (this is one of the reason why I split the logic into two passes). I can document it more explicitly that the order between those two is important.

Reply:
Having this specific ordering constraint between the passes is not really a good argument for having two passes to begin with... I think it argues more for having a single pass that performs all steps in the correct order.

Reply:
The main reason for having two passes is that I need IDs to be regenerated between them (see comment in processing manager)

Reply:
For this purpose I added the `localId` thing. You can run `IdReassingment.newInstance().run(<func>)` to reassign all local ids from within the pass.
```Java
processingStepOne(func);
IdReassingment.newInstance().run(func); // Reassign ids.
processingStepTwo(func);
```